### PR TITLE
Issue #3048: ParenPad now correctly evaluates precedence parens

### PR DESCRIPTION
**Pull request by subkrish**
Issue #3048

### CODE REVIEW

**Comment by rnveach** [comment]
TC and codecov errors can be ignored for now.

Taken from https://github.com/checkstyle/checkstyle/issues/3048:
> skip completely DOT sub tree when user define METHOD_CALL tag
>  add new token to Check, METHOD_CALL should skip validation non its parentheses.

@subkrish These items seem to be missing. Current code just adds `DOT` token.
It does fix issue for `METHOD_CALL` seen in removals for `testConfigureTokens` because `processExpression` tells it to.
Is this the reason?

**Comment by subkrish** [comment]
Yes, `processExpression` takes care of it. Also `METHOD_CALL` skips validation on its parentheses whenever there is an expression within the parentheses. I was trying to point it out in a comment made in the issue, when a warning was raised because of the `METHOD_CALL` parentheses. I wasn't able to understand the requirement then, but now I have got it.

**Comment by subkrish** [comment]
Here is diff report: https://subkrish.github.io/

No diff is present after the change. Also hibernate-search has been added.

**Comment by rnveach** [comment]
> No diff is present after the change.

@subkrish Correct me if I am wrong, but how can this be?
We have differences in violations in our own test files for this PR at https://github.com/checkstyle/checkstyle/pull/4109/files#diff-1e86eeef6e6d0c2cbd6a8a1df89be6f7L240 .
I would expect similar changes in regression.
Am I missing something?

**Comment by subkrish** [comment]
> We have differences in violations in our own test files for this PR

@rnveach To be honest, I wasn't to sure about the reports either just because the reason you mentioned but I was caught up in writing my proposal and couldn't get back to it.

I was actually thinking about running the reports again. Before I do I would like to just make sure that in `my_check.xml` when I add `ParenPadCheck` do I also add its property i.e. `<property name="tokens" value="METHOD_CALL"/>` and `<property name="option" value="space"/>` because this is what I was initially testing on?

**Comment by rnveach** [comment]
@subkrish I think I understand the differences now.
Since default module is using all tokens, it is processing everything and giving us all violations. The reason there are no differences is because some other token is picking up the violations that this fix is dropping. This is because we are doing a deep scan by `processExpression`.

Differences can be seen, if specific `tokens` is supplied for the tokens involved in this fix. We would have to run regression on each token to see who is adding new violations.
I re-ran 2 reports just to confirm your report's findings.
Default module: http://rveach.no-ip.org/checkstyle/regression/reports/12/ (no differences)
`METHOD_CALL` only: http://rveach.no-ip.org/checkstyle/regression/reports/13/ (differences)

**Comment by subkrish** [comment]
Like you have suggested, I had run the diff on a few projects and yes there were differences. I used `<property name="tokens" value="METHOD_CALL"/>`
`<property name="option" value="space"/>`
for `ParenPadCheck`.

`METHOD_CALL` and `space` : https://subkrish.github.io/
> We would have to run regression on each token to see who is adding new violations

Basically I would have to run it on `DOT` and `METHOD_CALL` right? I don't think other tokens are getting affected.

**Comment by rnveach** [comment]
> Basically I would have to run it on DOT and METHOD_CALL right?

As long as the numbers (additions and subtractions) even out to zero, yes.
If your change is affecting other tokens, then we would specifically need to look at those.

**Comment by subkrish** [comment]
There is an issue I am facing while checking `ParenPad` for `DOT`. Since `DOT` is not a token for `ParenPad` in my master branch it is throwing an error.

`Token "DOT" was not found in Acceptable tokens list in check com.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck`

How do I go about this?

**Comment by subkrish** [comment]
The final updated report that was run on `METHOD_CALL`, `DOT` and `EXPR` :  https://subkrish.github.io/.

Everything evens out between these 3 tokens.

**Comment by rnveach** [comment]
@romani Because of https://github.com/checkstyle/checkstyle/pull/4109#issuecomment-291590567 , I think we should consider dropping the deep scan and adding any tokens that get missed out.
Having 0 changes for a new token when specific tokens do see a difference is very weird.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputParenPad.java [hunk 1]
```diff
@@ -131,4 +131,28 @@ public boolean bar(boolean a) {
...
+        String s = "test";
+        Object o = s;
+        ((String)o).length();
+        ( (String)o ).length();
+    }
+
+    public void crisRon() {
+        Object leo = "messi";
+        Object ibra = leo;
+        ((String)leo).compareTo( (String)ibra );
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputParenPad.java]: please add a bit more complicated test cases in inputs, you can take ideas from your report.
combined cases and :
`Boolean value = (Boolean) ( (TwoWayFieldBridge) fieldBridge ).get( field.name(), document );`
`Query query = ( (QueryWrapperFilter) luceneFilter ).getQuery();`
`assertThat( ( (Number) projection[0] ).intValue() ).describedAs( "fieldNotInMapping" ).isEqualTo( 42 );`
.... please add about 10 different and complicated cases.

**Comment by rnveach** [comment]
> yes, please create new issue on this. Please make a comment in original issue that issue will be partly fixed and final fix is separate issue.

Done.

**Comment by romani** [comment]
@subkrish , we expected additional changes from you, please continue this PR.

**Comment by subkrish** [comment]
@romani @rnveach I have made the changes. Please review.

**Comment by rnveach** [comment]
@subkrish Please keep your branches up to date with master so we can avoid any conflicts after a merge.
Please rebase your branch.

**Comment by subkrish** [comment]
@rnveach I have rebased my branch. Please review.

**Pull request merged**